beacon-chain/sync/subscriber.go

```diff
@@ -41,55 +42,67 @@ func (s *Service) noopValidator(_ context.Context, _ peer.ID, msg *pubsub.Messag
 }
 
 // Register PubSub subscribers
-func (s *Service) registerSubscribers() {
+func (s *Service) registerSubscribers(epoch types.Epoch, digest [4]byte) {
 	s.subscribe(
 		p2p.BlockSubnetTopicFormat,
 		s.validateBeaconBlockPubSub,
 		s.beaconBlockSubscriber,
+		digest,
 	)
 	s.subscribe(
 		p2p.AggregateAndProofSubnetTopicFormat,
 		s.validateAggregateAndProof,
 		s.beaconAggregateProofSubscriber,
+		digest,
 	)
 	s.subscribe(
 		p2p.ExitSubnetTopicFormat,
 		s.validateVoluntaryExit,
 		s.voluntaryExitSubscriber,
+		digest,
 	)
 	s.subscribe(
 		p2p.ProposerSlashingSubnetTopicFormat,
 		s.validateProposerSlashing,
 		s.proposerSlashingSubscriber,
+		digest,
 	)
 	s.subscribe(
 		p2p.AttesterSlashingSubnetTopicFormat,
 		s.validateAttesterSlashing,
 		s.attesterSlashingSubscriber,
+		digest,
 	)
 	if flags.Get().SubscribeToAllSubnets {
 		s.subscribeStaticWithSubnets(
-			"/eth2/%x/beacon_attestation_%d",
+			p2p.AttestationSubnetTopicFormat,
 			s.validateCommitteeIndexBeaconAttestation,   /* validator */
 			s.committeeIndexBeaconAttestationSubscriber, /* message handler */
+			digest,
 		)
 	} else {
 		s.subscribeDynamicWithSubnets(
-			"/eth2/%x/beacon_attestation_%d",
+			p2p.AttestationSubnetTopicFormat,
 			s.validateCommitteeIndexBeaconAttestation,   /* validator */
 			s.committeeIndexBeaconAttestationSubscriber, /* message handler */
+			digest,
 		)
 	}
 }
 
 // subscribe to a given topic with a given validator and subscription handler.
 // The base protobuf message is used to initialize new messages for decoding.
-func (s *Service) subscribe(topic string, validator pubsub.ValidatorEx, handle subHandler) *pubsub.Subscription {
-	base := p2p.GossipTopicMappings(topic, 0)
+func (s *Service) subscribe(topic string, validator pubsub.ValidatorEx, handle subHandler, digest [4]byte) *pubsub.Subscription {
+	genRoot := s.cfg.Chain.GenesisValidatorRoot()
+	_, e, err := p2putils.RetrieveForkDataFromDigest(digest, genRoot[:])
+	if err != nil {
+		panic(err)
```

Comment:
panic? Can't we change the signature to `(*pubsub.Subscription, error)`?

Reply:
It is a panic because it is an impossible condition, if we hit this there is no way to continue onwards as the node wouldn't be subscribed to any topic.

Reply:
This is a good comment to add IMO. Any future reader will immediately question the panic otherwise @nisdas